src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_BuildSynchronization.cs

```diff
@@ -59,27 +52,41 @@ public async Task SynchronizeWithBuildAsync(ImmutableDictionary<ProjectId, Immut
                         continue;
                     }
 
-                    // REVIEW: do build diagnostics include suppressed diagnostics?
                     var stateSets = _stateManager.CreateBuildOnlyProjectStateSet(project);
+                    var newResult = CreateAnalysisResults(project, stateSets, diagnostics);
 
-                    // We load data since we don't know right version.
-                    var oldAnalysisData = await ProjectAnalysisData.CreateAsync(PersistentStorageService, project, stateSets, avoidLoadingData: false, cancellationToken).ConfigureAwait(false);
-                    var newResult = CreateAnalysisResults(project, stateSets, oldAnalysisData, diagnostics);
-
+                    // PERF: Save the diagnostics into in-memory cache on the main thread.
+                    //       Saving them into persistent storage is expensive, so we invoke that operation on a separate task queue
+                    //       to ensure faster error list refresh.
                     foreach (var stateSet in stateSets)
                     {
                         cancellationToken.ThrowIfCancellationRequested();
 
                         var state = stateSet.GetOrCreateProjectState(project.Id);
                         var result = GetResultOrEmpty(newResult, stateSet.Analyzer, project.Id, VersionStamp.Default);
-                        await state.SaveAsync(PersistentStorageService, project, result).ConfigureAwait(false);
+
+                        // Save into in-memory cache.
+                        await state.SaveInMemoryCacheAsync(project, result).ConfigureAwait(false);
+
+                        // Save into persistent storage on a separate post-build task queue.
+                        _ = postBuildTaskQueue.ScheduleTask(
+                                nameof(SynchronizeWithBuildAsync),
+                                async () => await state.SaveAsync(PersistentStorageService, project, result).ConfigureAwait(false),
+                                cancellationToken);
                     }
 
-                    // REVIEW: this won't handle active files. might need to tweak it later.
-                    RaiseProjectDiagnosticsIfNeeded(project, stateSets, oldAnalysisData.Result, newResult);
+                    // Raise diagnostic updated events after the new diagnostics have been stored into the in-memory cache.
+                    if (diagnostics.IsEmpty)
+                    {
+                        ClearAllDiagnostics(stateSets, projectId);
+                    }
+                    else
+                    {
+                        RaiseProjectDiagnosticsIfNeeded(project, stateSets, newResult);
+                    }
```

Comment:
fwiw, this condition just reads really weird.  i.e. one branch uses 'diagnostics' as the data, the other uses 'newResult'.  and we process these two separate pieces of data here to determine what to do.  I would expect a general processing routine to just handle project/staeSets/newResult and decide elsewhere what to do.  

no need to change this... but it just smells funny to me.

Reply:
This was needed because newResult will always have entries for each diagnostic analyzer, even for the case where there are no diagnostics. For that case, we are only interesting in clearing all diagnostics, but do not have a cheap way to check that from `newResult`, hence this check.